File: internal/tools/attachments.go
package tools

import (
	"context"
	"errors"
	"fmt"
	"io/fs"
	"log"
	"os"
	"path/filepath"
	"runtime"
	"strings"

	"github.com/dsaiko/edookit-mcp/internal/client"
)

// DownloadResult is what DownloadAttachments returns: one entry per
// attachment Edookit listed on the message, ordered as the message presents
// them. A non-empty Error on an entry means that single download failed
// (other entries continued); a top-level error from DownloadAttachments
// itself means the call couldn't even start (bad ID, can't create the
// destination dir, etc.).
type DownloadResult struct {
	MessageID string           `json:"message_id"`
	Directory string           `json:"directory"`
	Files     []DownloadedFile `json:"files"`
}

// DownloadedFile is one attachment after we attempted to fetch it.
// Bytes is the actual on-disk size. Skipped is true when an existing file
// at the destination path was left alone (only possible with
// Overwrite=false — the call doesn't compare sizes, just presence). Name
// is Edookit's original filename even when Path was disambiguated by a
// suffix to avoid colliding with an earlier attachment in the same call.
type DownloadedFile struct {
	Name    string `json:"name"`
	Path    string `json:"path"`
	Bytes   int64  `json:"bytes"`
	Skipped bool   `json:"skipped,omitempty"`
	Error   string `json:"error,omitempty"`
}

// DownloadOptions controls DownloadAttachments. DestDir is the target
// directory; an empty value resolves to <os.TempDir>/edookit-mcp/<message-id>,
// which lands in the OS-appropriate temp location (/tmp on Unix,
// %TMP%\... on Windows). Tilde-expansion is applied to non-empty paths,
// and the directory tree is created on demand with 0700 perms (the files
// inside come from a private school account).
type DownloadOptions struct {
	DestDir   string
	Overwrite bool // if false, an existing file at the destination path is left untouched (no size check — we have no Content-Length up-front)
}

// DownloadAttachments resolves the given message ID, downloads every
// (non-trashed) attachment via the authenticated session, and writes each
// to DestDir/<original-filename>. Returns a per-file outcome. Partial
// failures don't abort the loop — each attachment's outcome is captured in
// its DownloadedFile entry so a single broken file doesn't lose the rest.
func DownloadAttachments(ctx context.Context, cli *client.Client, messageID string, opts DownloadOptions) (*DownloadResult, error) {
	msg, err := GetMessage(ctx, cli, messageID)
	if err != nil {
		return nil, err
	}

	destDir, err := resolveDestDir(opts.DestDir, msg.ID)
	if err != nil {
		return nil, err
	}
	if err := os.MkdirAll(destDir, 0o700); err != nil {
		return nil, fmt.Errorf("create destination dir %s: %w", destDir, err)
	}

	res := &DownloadResult{
		MessageID: msg.ID,
		Directory: destDir,
		Files:     make([]DownloadedFile, 0, len(msg.Attachments)),
	}

	// usedNames tracks final base names already taken inside this call so a
	// second attachment with the same Edookit-supplied filename gets a
	// "-2" / "-3" suffix instead of being silently skipped (Overwrite=false)
	// or clobbering the first (Overwrite=true).
	usedNames := make(map[string]int, len(msg.Attachments))
	for _, a := range msg.Attachments {
		entry := downloadOne(ctx, cli, a, destDir, opts.Overwrite, usedNames)
		res.Files = append(res.Files, entry)
	}
	return res, nil
}

// downloadOne handles a single attachment end-to-end. Returns the result
// entry; never panics. Logs at debug-level for the operator. usedNames is
// shared across the whole DownloadAttachments call so duplicate filenames
// within one message get disambiguated with a "-N" suffix instead of
// silently colliding.
func downloadOne(ctx context.Context, cli *client.Client, a Attachment, destDir string, overwrite bool, usedNames map[string]int) DownloadedFile {
	out := DownloadedFile{Name: a.Name}

	if errMsg := validateAttachment(a); errMsg != "" {
		out.Error = errMsg
		return out
	}

	dst, errMsg := planDestination(a.Name, destDir, usedNames)
	if errMsg != "" {
		out.Error = errMsg
		return out
	}
	out.Path = dst

	return streamToDest(ctx, cli, a.URL, destDir, dst, overwrite, out)
}

// validateAttachment performs the cheap up-front checks that don't depend
// on the destination directory: a present download URL, and a base name
// that isn't a directory-traversal sentinel or Windows-reserved. Returns
// an error message string (empty = OK) so the caller can stuff it into
// the result entry without an error allocation.
func validateAttachment(a Attachment) string {
	// Server-supplied URL validation: an empty / whitespace-only link
	// would otherwise resolve through Client.resolve("") to the base URL
	// itself, so we'd happily save the Edookit landing page as the
	// "attachment". Fail per-file instead.
	if strings.TrimSpace(a.URL) == "" {
		return "attachment has no download URL"
	}
	safeName := filepath.Base(a.Name)
	if safeName == "" || safeName == "." || safeName == ".." || safeName == string(os.PathSeparator) {
		return fmt.Sprintf("attachment name %q is not safe to use as a filename", a.Name)
	}
	if reason := windowsUnsafeName(safeName); reason != "" {
		return fmt.Sprintf("attachment name %q rejected: %s", a.Name, reason)
	}
	return ""
}

// planDestination computes the final on-disk path for an attachment,
// applying within-call filename deduplication and verifying the result
// stays inside destDir. Returns (dst, "") on success or ("", errMsg) if
// the name can't be placed safely. Mutates usedNames to claim the chosen
// filename so subsequent attachments in the same call get disambiguated.
func planDestination(name, destDir string, usedNames map[string]int) (dst, errMsg string) {
	safeName := filepath.Base(name)
	// Within-call collision: if a previous attachment in this same message
	// already claimed this filename, bump a "-N" suffix until free. The
	// skip-if-exists check in the caller handles previous-run files
	// separately (that's pre-existing-file semantics, not dedup), so this
	// only looks at usedNames.
	safeName = uniqueFilename(safeName, usedNames)
	usedNames[strings.ToLower(safeName)]++

	dst = filepath.Join(destDir, safeName)

	// Path-traversal defense, step 2: even after filepath.Base, a Windows
	// volume-rooted name like "C:foo" can survive Join in ways that escape
	// destDir. Verify dst is inside destDir via filepath.Rel; a result
	// starting with ".." means containment was broken.
	rel, relErr := filepath.Rel(destDir, dst)
	if relErr != nil || strings.HasPrefix(rel, "..") || rel == ".." {
		return "", fmt.Sprintf("attachment name %q would escape destination dir", name)
	}
	return dst, ""
}

// streamToDest writes the body of url to a unique temp file in destDir and
// then commits it to dst. The download always lands in a hidden .part temp
// first, so:
//   - a reader never sees a half-written or 0-byte file at dst, and
//   - a failed/aborted download leaves at most an orphan temp, never a
//     partial final file.
//
// The commit step differs by mode:
//   - overwrite=true: atomicRename replaces any existing dst.
//   - no-overwrite: os.Link the completed temp to dst. The link only appears
//     once it points at the fully-downloaded file (no placeholder window),
//     and it fails with fs.ErrExist if dst already exists → keep it (Skipped).
//     This makes the "keep existing files" guarantee race-free without ever
//     exposing dst before it holds complete content: a concurrent caller
//     either wins the link or sees a finished file, never an empty reservation.
//     (os.Link needs dst and temp on the same filesystem, which holds since
//     both live in destDir; it is unsupported on a few filesystems like FAT.)
//
// `out` is the already-populated result entry — we layer Bytes / Error
// onto it and return.
func streamToDest(ctx context.Context, cli *client.Client, url, destDir, dst string, overwrite bool, out DownloadedFile) DownloadedFile {
	// Fast path: skip an already-present file without downloading it. This is
	// purely an optimization — correctness for the race where dst appears
	// mid-download is still enforced by the os.Link commit below, which fails
	// with ErrExist rather than clobbering.
	if !overwrite {
		if info, err := os.Stat(dst); err == nil {
			out.Skipped = true
			out.Bytes = info.Size()
			return out
		}
	}

	tmpf, err := os.CreateTemp(destDir, ".edookit-download-*.part")
	if err != nil {
		out.Error = fmt.Sprintf("create temp in %s: %v", destDir, err)
		return out
	}
	tmpPath := tmpf.Name()
	defer func() {
		if tmpPath != "" {
			_ = os.Remove(tmpPath)
		}
	}()
	if err := tmpf.Chmod(0o600); err != nil {
		_ = tmpf.Close()
		out.Error = fmt.Sprintf("chmod %s: %v", tmpPath, err)
		return out
	}

	n, copyErr := cli.GetTo(ctx, url, tmpf)
	closeErr := tmpf.Close()
	if copyErr != nil {
		out.Error = fmt.Sprintf("download: %v", copyErr)
		return out
	}
	if closeErr != nil {
		out.Error = fmt.Sprintf("close: %v", closeErr)
		return out
	}

	if !overwrite {
		// Hard-link the finished temp into place. dst becomes visible only as
		// a complete file; an existing dst makes Link fail with ErrExist so we
		// keep it. The temp name is dropped by the deferred cleanup either way.
		if err := os.Link(tmpPath, dst); err != nil {
			if errors.Is(err, fs.ErrExist) {
				out.Skipped = true
				if info, statErr := os.Stat(dst); statErr == nil {
					out.Bytes = info.Size()
				}
				return out
			}
			out.Error = fmt.Sprintf("link %s -> %s: %v", tmpPath, dst, err)
			return out
		}
		out.Bytes = n
		log.Printf("[tools] downloaded %d bytes -> %s", n, dst)
		return out
	}

	// overwrite=true: replace any existing dst atomically.
	if err := atomicRename(tmpPath, dst); err != nil {
		out.Error = fmt.Sprintf("rename %s -> %s: %v", tmpPath, dst, err)
		return out
	}
	tmpPath = "" // rename consumed the temp; deferred cleanup must not remove dst
	out.Bytes = n
	log.Printf("[tools] downloaded %d bytes -> %s", n, dst)
	return out
}

// uniqueFilename returns base if not already claimed within this call, or
// "<stem>-N.<ext>" with the smallest N >= 2 that is free. Operates on the
// usedNames map only — pre-existing files on disk are handled by the
// skip-if-exists / overwrite policy in downloadOne, not here.
//
// Keys in usedNames are normalized to lowercase so "Report.pdf" and
// "report.pdf" collide as they would on case-insensitive filesystems
// (macOS APFS default, Windows NTFS). Doing it unconditionally is harmless
// on case-sensitive ones (Linux ext4) — at worst it picks "-2" when the
// strict semantics would have allowed both names to coexist, which is a
// negligible loss of distinguishability vs. the silent corruption it
// prevents on case-insensitive volumes.
func uniqueFilename(base string, usedNames map[string]int) string {
	if _, taken := usedNames[strings.ToLower(base)]; !taken {
		return base
	}
	ext := filepath.Ext(base)
	stem := strings.TrimSuffix(base, ext)
	for n := 2; ; n++ {
		candidate := fmt.Sprintf("%s-%d%s", stem, n, ext)
		if _, taken := usedNames[strings.ToLower(candidate)]; !taken {
			return candidate
		}
	}
}

// resolveDestDir applies the default-path policy and expands a leading
// tilde. Default (when raw is empty): <os.TempDir>/edookit-mcp/<message-id>.
// os.TempDir returns the OS-appropriate temp directory (/tmp on Unix,
// %TMP%\... on Windows, $TMPDIR on macOS) so the default is portable and
// gets garbage-collected by the OS — users who want persistence pass an
// explicit path.
//
// Relative paths are rejected: the MCP server's cwd is whatever started
// the host application (Claude Desktop, Code, ChatGPT, …) which is
// unpredictable and would land files in a surprising place. The README
// contract documents this — caller must pass an absolute path or one
// starting with "~/".
func resolveDestDir(raw, msgID string) (string, error) {
	if raw == "" {
		return filepath.Join(os.TempDir(), "edookit-mcp", msgID), nil
	}
	if strings.HasPrefix(raw, "~/") || raw == "~" {
		home, err := os.UserHomeDir()
		if err != nil {
			return "", fmt.Errorf("expand ~ in %q: %w", raw, err)
		}
		if raw == "~" {
			return home, nil
		}
		return filepath.Join(home, raw[2:]), nil
	}
	if !filepath.IsAbs(raw) {
		return "", fmt.Errorf("destination_dir %q must be absolute or start with ~/ (the MCP server's cwd is not a stable anchor)", raw)
	}
	return raw, nil
}

// windowsReservedBasenames is the set of DOS device names Windows refuses
// to create as regular files, applied case-insensitively and regardless
// of extension (CON.txt is reserved too). Sourced from
// https://learn.microsoft.com/windows/win32/fileio/naming-a-file.
var windowsReservedBasenames = map[string]bool{
	"CON": true, "PRN": true, "AUX": true, "NUL": true,
	"COM0": true, "COM1": true, "COM2": true, "COM3": true, "COM4": true,
	"COM5": true, "COM6": true, "COM7": true, "COM8": true, "COM9": true,
	"LPT0": true, "LPT1": true, "LPT2": true, "LPT3": true, "LPT4": true,
	"LPT5": true, "LPT6": true, "LPT7": true, "LPT8": true, "LPT9": true,
}

// windowsUnsafeName returns a non-empty reason string when name would
// misbehave on Windows. Catches:
//
//   - Reserved characters: ":" (creates an NTFS Alternate Data Stream
//     inside destDir instead of a normal file), plus < > " | ? * and
//     control bytes (0-31).
//   - Reserved DOS device basenames: CON, PRN, AUX, NUL, COM0-9, LPT0-9
//     — case-insensitive, with or without an extension.
//   - Names ending in space or '.' (Windows silently strips them and
//     creates a file under a name the user can't reliably re-open).
//
// All checks per https://learn.microsoft.com/windows/win32/fileio/naming-a-file.
// Always returns "" on non-Windows GOOS so otherwise-valid Unix filenames
// (containing "abc:def" or named "CON") aren't blocked.
func windowsUnsafeName(name string) string {
	if runtime.GOOS != "windows" {
		return ""
	}
	for _, r := range name {
		if r < 32 {
			return "contains a control character (Windows-reserved)"
		}
		switch r {
		case ':':
			return "contains ':' (would create an NTFS Alternate Data Stream on Windows)"
		case '<', '>', '"', '|', '?', '*':
			return fmt.Sprintf("contains Windows-reserved character %q", r)
		}
	}
	// Trailing space or dot: Windows quietly strips them, so the actual
	// on-disk name doesn't match what was requested and re-opening by
	// the requested name fails.
	if last := name[len(name)-1]; last == ' ' || last == '.' {
		return "ends with space or '.' (Windows silently strips trailing space/dot)"
	}
	// DOS device basename: reserved regardless of extension. Strip
	// extension and compare uppercase against the well-known list.
	stem := name
	if dot := strings.LastIndexByte(name, '.'); dot >= 0 {
		stem = name[:dot]
	}
	if windowsReservedBasenames[strings.ToUpper(stem)] {
		return fmt.Sprintf("basename %q is a reserved Windows device name", stem)
	}
	return ""
}

// atomicRename os.Renames tmp to dst with one fallback: on older Windows
// or restrictive ACL configs, Rename can return fs.ErrExist when dst
// already exists (overwrite=true case) — even though Go 1.5+ MoveFileEx
// is supposed to replace it. In that case only, remove dst and retry.
// Other errors (perms, I/O, cross-device, disk full) propagate without
// touching dst so the user's previous file is preserved. Same pattern
// internal/client/cookie_store.go uses.
func atomicRename(tmp, dst string) error {
	err := os.Rename(tmp, dst)
	if err == nil {
		return nil
	}
	if !errors.Is(err, fs.ErrExist) {
		return err
	}
	if removeErr := os.Remove(dst); removeErr != nil && !errors.Is(removeErr, fs.ErrNotExist) {
		return fmt.Errorf("remove existing dst before rename retry: %w (original rename error: %w)", removeErr, err)
	}
	return os.Rename(tmp, dst)
}

File: src/tools/attachments.rs
//! `download_attachments` — streams each attachment of a message into a local
//! directory. Port of Go's `internal/tools/attachments.go`.

use std::collections::HashMap;
use std::path::{Path, PathBuf};

use anyhow::{Context, anyhow, bail};
use serde::Serialize;

use super::message::{Attachment, get_message};
use crate::client::Client;

/// One entry per attachment Edookit listed, ordered as the message presents
/// them. A non-empty `error` means that single download failed (others
/// continued); a top-level error means the call couldn't even start.
#[derive(Debug, Clone, Serialize)]
pub struct DownloadResult {
    pub message_id: String,
    pub directory: String,
    pub files: Vec<DownloadedFile>,
}

#[derive(Debug, Clone, Default, Serialize)]
pub struct DownloadedFile {
    pub name: String,
    pub path: String,
    pub bytes: i64,
    #[serde(skip_serializing_if = "is_false")]
    pub skipped: bool,
    #[serde(skip_serializing_if = "String::is_empty")]
    pub error: String,
}

fn is_false(b: &bool) -> bool {
    !*b
}

#[derive(Debug, Default, Clone)]
pub struct DownloadOptions {
    pub dest_dir: String,
    /// If false, an existing file at the destination is left untouched.
    pub overwrite: bool,
}

/// Resolves the message ID, downloads every (non-trashed) attachment via the
/// authenticated session, and writes each to `dest_dir/<original-filename>`.
/// Partial failures don't abort the loop.
pub async fn download_attachments(
    cli: &Client,
    message_id: &str,
    opts: DownloadOptions,
) -> anyhow::Result<DownloadResult> {
    let msg = get_message(cli, message_id).await?;

    let dest_dir = resolve_dest_dir(&opts.dest_dir, &msg.id)?;
    create_dir_secure(&dest_dir).with_context(|| format!("create destination dir {}", dest_dir.display()))?;

    let mut res = DownloadResult {
        message_id: msg.id.clone(),
        directory: dest_dir.display().to_string(),
        files: Vec::with_capacity(msg.attachments.len()),
    };

    // Tracks final base names taken inside this call so a duplicate filename
    // gets a "-2"/"-3" suffix instead of being skipped or clobbered.
    let mut used_names: HashMap<String, ()> = HashMap::new();
    for a in &msg.attachments {
        let entry = download_one(cli, a, &dest_dir, opts.overwrite, &mut used_names).await;
        res.files.push(entry);
    }
    Ok(res)
}

async fn download_one(
    cli: &Client,
    a: &Attachment,
    dest_dir: &Path,
    overwrite: bool,
    used_names: &mut HashMap<String, ()>,
) -> DownloadedFile {
    let mut out = DownloadedFile {
        name: a.name.clone(),
        ..Default::default()
    };
    if let Some(reason) = validate_attachment(a) {
        out.error = reason;
        return out;
    }
    let dst = match plan_destination(&a.name, dest_dir, used_names) {
        Ok(d) => d,
        Err(reason) => {
            out.error = reason;
            return out;
        }
    };
    out.path = dst.display().to_string();
    stream_to_dest(cli, &a.url, dest_dir, &dst, overwrite, out).await
}

/// Cheap up-front checks not depending on the destination dir: a present URL,
/// and a base name that isn't a traversal sentinel or Windows-reserved.
fn validate_attachment(a: &Attachment) -> Option<String> {
    if a.url.trim().is_empty() {
        return Some("attachment has no download URL".to_string());
    }
    let safe = base_name(&a.name);
    match safe.as_deref() {
        None | Some("") | Some(".") | Some("..") => {
            Some(format!("attachment name {:?} is not safe to use as a filename", a.name))
        }
        Some(name) => windows_unsafe_name(name).map(|r| format!("attachment name {:?} rejected: {}", a.name, r)),
    }
}

/// Computes the final on-disk path, applying within-call dedup and verifying
/// the result stays inside `dest_dir`.
fn plan_destination(
    name: &str,
    dest_dir: &Path,
    used_names: &mut HashMap<String, ()>,
) -> Result<PathBuf, String> {
    let base = base_name(name).ok_or_else(|| format!("attachment name {name:?} is not safe to use as a filename"))?;
    let safe = unique_filename(&base, used_names);
    used_names.insert(safe.to_lowercase(), ());

    let dst = dest_dir.join(&safe);
    // Containment: the chosen name is a single component, so dst's parent must
    // be exactly dest_dir. (Defends against volume-rooted Windows names too.)
    if dst.parent() != Some(dest_dir) {
        return Err(format!("attachment name {name:?} would escape destination dir"));
    }
    Ok(dst)
}

/// Writes the body of `url` to a temp file in `dest_dir`, then commits it to
/// `dst`: `persist_noclobber` (no-overwrite, race-free — fails if dst exists)
/// or `persist` (overwrite, atomic replace). A reader never sees a partial file.
async fn stream_to_dest(
    cli: &Client,
    url: &str,
    dest_dir: &Path,
    dst: &Path,
    overwrite: bool,
    mut out: DownloadedFile,
) -> DownloadedFile {
    // Fast path: skip an already-present file without downloading. Correctness
    // for the mid-download race is still enforced by persist_noclobber below.
    if !overwrite && let Ok(meta) = std::fs::metadata(dst) {
        out.skipped = true;
        out.bytes = meta.len() as i64;
        return out;
    }

    let mut tmp = match tempfile::Builder::new().prefix(".edookit-download-").suffix(".part").tempfile_in(dest_dir) {
        Ok(t) => t,
        Err(e) => {
            out.error = format!("create temp in {}: {e}", dest_dir.display());
            return out;
        }
    };
    #[cfg(unix)]
    {
        use std::os::unix::fs::PermissionsExt;
        if let Err(e) = tmp.as_file().set_permissions(std::fs::Permissions::from_mode(0o600)) {
            out.error = format!("chmod temp: {e}");
            return out;
        }
    }

    let n = match cli.get_to(url, tmp.as_file_mut()).await {
        Ok(n) => n,
        Err(e) => {
            out.error = format!("download: {e}");
            return out;
        }
    };

    if !overwrite {
        // persist_noclobber: dst becomes visible only as a complete file; an
        // existing dst makes it fail with AlreadyExists, so we keep it.
        match tmp.persist_noclobber(dst) {
            Ok(_) => {
                out.bytes = n as i64;
                tracing::info!("downloaded {n} bytes -> {}", dst.display());
            }
            Err(e) if e.error.kind() == std::io::ErrorKind::AlreadyExists => {
                out.skipped = true;
                if let Ok(meta) = std::fs::metadata(dst) {
                    out.bytes = meta.len() as i64;
                }
            }
            Err(e) => {
                out.error = format!("link -> {}: {}", dst.display(), e.error);
            }
        }
        return out;
    }

    // overwrite: atomic replace.
    match tmp.persist(dst) {
        Ok(_) => {
            out.bytes = n as i64;
            tracing::info!("downloaded {n} bytes -> {}", dst.display());
        }
        Err(e) => {
            out.error = format!("rename -> {}: {}", dst.display(), e.error);
        }
    }
    out
}

/// Returns `base` if free within this call, else "<stem>-N.<ext>" with the
/// smallest free N >= 2. Keys are lowercased so case-insensitive filesystems
/// don't silently alias.
fn unique_filename(base: &str, used_names: &HashMap<String, ()>) -> String {
    if !used_names.contains_key(&base.to_lowercase()) {
        return base.to_string();
    }
    let p = Path::new(base);
    let ext = p.extension().and_then(|e| e.to_str()).map(|e| format!(".{e}")).unwrap_or_default();
    let stem = base.strip_suffix(&ext).unwrap_or(base);
    let mut n = 2;
    loop {
        let candidate = format!("{stem}-{n}{ext}");
        if !used_names.contains_key(&candidate.to_lowercase()) {
            return candidate;
        }
        n += 1;
    }
}

/// `filepath.Base` equivalent: the final path component, or `None` for paths
/// with no usable file name (`""`, `"."`, `".."`, `"/"`).
fn base_name(name: &str) -> Option<String> {
    Path::new(name).file_name().and_then(|s| s.to_str()).map(|s| s.to_string())
}

/// Applies the default-path policy and expands a leading tilde. Empty →
/// `<temp-dir>/edookit-mcp/<message-id>`. Relative paths are rejected (the MCP
/// server's cwd is not a stable anchor).
fn resolve_dest_dir(raw: &str, msg_id: &str) -> anyhow::Result<PathBuf> {
    if raw.is_empty() {
        return Ok(std::env::temp_dir().join("edookit-mcp").join(msg_id));
    }
    if raw == "~" || raw.starts_with("~/") {
        let home = dirs::home_dir().ok_or_else(|| anyhow!("expand ~ in {raw:?}: no home dir"))?;
        return Ok(if raw == "~" { home } else { home.join(&raw[2..]) });
    }
    if !Path::new(raw).is_absolute() {
        bail!("destination_dir {raw:?} must be absolute or start with ~/ (the MCP server's cwd is not a stable anchor)");
    }
    Ok(PathBuf::from(raw))
}

#[cfg(unix)]
fn create_dir_secure(dir: &Path) -> std::io::Result<()> {
    use std::os::unix::fs::DirBuilderExt;
    std::fs::DirBuilder::new().recursive(true).mode(0o700).create(dir)
}
#[cfg(not(unix))]
fn create_dir_secure(dir: &Path) -> std::io::Result<()> {
    std::fs::create_dir_all(dir)
}

#[cfg(windows)]
fn windows_unsafe_name(name: &str) -> Option<String> {
    const RESERVED: &[&str] = &[
        "CON", "PRN", "AUX", "NUL", "COM0", "COM1", "COM2", "COM3", "COM4", "COM5", "COM6", "COM7",
        "COM8", "COM9", "LPT0", "LPT1", "LPT2", "LPT3", "LPT4", "LPT5", "LPT6", "LPT7", "LPT8", "LPT9",
    ];
    for c in name.chars() {
        if (c as u32) < 32 {
            return Some("contains a control character (Windows-reserved)".into());
        }
        match c {
            ':' => return Some("contains ':' (would create an NTFS Alternate Data Stream on Windows)".into()),
            '<' | '>' | '"' | '|' | '?' | '*' => return Some(format!("contains Windows-reserved character {c:?}")),
            _ => {}
        }
    }
    if let Some(last) = name.chars().last()
        && (last == ' ' || last == '.')
    {
        return Some("ends with space or '.' (Windows silently strips trailing space/dot)".into());
    }
    let stem = name.split_once('.').map(|(s, _)| s).unwrap_or(name);
    if RESERVED.contains(&stem.to_uppercase().as_str()) {
        return Some(format!("basename {stem:?} is a reserved Windows device name"));
    }
    None
}
#[cfg(not(windows))]
fn windows_unsafe_name(_name: &str) -> Option<String> {
    None
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::client::{Client, Config, LoginCookie, LoginFn};
    use std::sync::Arc;
    use std::time::Duration;
    use wiremock::matchers::{method, path as mpath, query_param};
    use wiremock::{Mock, MockServer, ResponseTemplate};

    #[test]
    fn resolve_dest_dir_policy() {
        // empty → temp/edookit-mcp/<id>
        let d = resolve_dest_dir("", "m-1").unwrap();
        assert!(d.ends_with("edookit-mcp/m-1"));
        // absolute passes through
        assert_eq!(resolve_dest_dir("/tmp/x", "m-1").unwrap(), PathBuf::from("/tmp/x"));
        // relative rejected
        assert!(resolve_dest_dir("rel/path", "m-1").is_err());
        // ~ expands
        assert!(resolve_dest_dir("~", "m-1").is_ok());
    }

    #[test]
    fn plan_destination_dedup_and_containment() {
        let dir = PathBuf::from("/dest");
        let mut used = HashMap::new();
        let a = plan_destination("report.pdf", &dir, &mut used).unwrap();
        assert_eq!(a, PathBuf::from("/dest/report.pdf"));
        // case-insensitive dedup: second "Report.PDF" gets -2
        let b = plan_destination("Report.PDF", &dir, &mut used).unwrap();
        assert_eq!(b, PathBuf::from("/dest/Report-2.PDF"));
    }

    #[test]
    fn traversal_name_collapses_to_basename() {
        let dir = PathBuf::from("/dest");
        let mut used = HashMap::new();
        // file_name of a traversal path is just the leaf, landing inside dest.
        let d = plan_destination("../../etc/passwd", &dir, &mut used).unwrap();
        assert_eq!(d, PathBuf::from("/dest/passwd"));
    }

    #[test]
    fn validate_attachment_rejects_empty_url() {
        let a = Attachment { id: "1".into(), name: "x.txt".into(), url: "".into(), date: String::new() };
        assert!(validate_attachment(&a).is_some());
    }

    #[test]
    fn windows_unsafe_name_is_noop_off_windows() {
        // On non-Windows, otherwise-valid Unix names are allowed.
        assert!(windows_unsafe_name("CON").is_none() || cfg!(windows));
    }

    fn build_client(uri: &str) -> Client {
        let login_fn: LoginFn = Arc::new(|| Box::pin(async { Ok(vec![LoginCookie::new("X-EdooAuthToken", "tok")]) }));
        let mut cfg = Config::new(uri, "u", "p");
        cfg.retry_base_delay = Duration::from_millis(1);
        cfg.login_fn = Some(login_fn);
        Client::new(cfg).unwrap()
    }

    fn message_edit_with_attachment(download_url: &str) -> serde_json::Value {
        serde_json::json!({
            "authenticated": true,
            "components": {"workspace": [{
                "DOMTarget": "__lc_Form_Message",
                "data": {"__form_panel_main": [{"items": [
                    {"name": "name", "val": "Subj"},
                    {"name": "object_status", "val": "<span>Publikováno</span>"},
                    {"name": "description__editor", "readValue": "<p>body</p>"}
                ]}]}
            }, {
                "DOMTarget": "__lc_Fileviewer_Slave_datatemplate_message",
                "data": {"data": [
                    {"id": "1@1", "name": "schedule.pdf", "link": download_url, "date": 0, "trashed": false}
                ]}
            }]}
        })
    }

    #[tokio::test]
    async fn downloads_attachment_to_dir() {
        let server = MockServer::start().await;
        Mock::given(method("GET")).and(mpath("/")).respond_with(ResponseTemplate::new(200)).mount(&server).await;
        let dl_url = format!("{}/handler/download/file1", server.uri());
        Mock::given(method("GET"))
            .and(mpath("/handler/page/message-edit"))
            .and(query_param("__index", "1"))
            .respond_with(ResponseTemplate::new(200).set_body_json(message_edit_with_attachment(&dl_url)))
            .mount(&server)
            .await;
        Mock::given(method("GET"))
            .and(mpath("/handler/download/file1"))
            .respond_with(
                ResponseTemplate::new(200)
                    .insert_header("Content-Type", "application/octet-stream")
                    .set_body_bytes(b"%PDF-1.4 hello".to_vec()),
            )
            .mount(&server)
            .await;

        let tmp = tempfile::tempdir().unwrap();
        let cli = build_client(&server.uri());
        let res = download_attachments(
            &cli,
            "m-1",
            DownloadOptions { dest_dir: tmp.path().display().to_string(), overwrite: false },
        )
        .await
        .unwrap();

        assert_eq!(res.files.len(), 1);
        assert_eq!(res.files[0].name, "schedule.pdf");
        assert_eq!(res.files[0].error, "");
        assert_eq!(res.files[0].bytes, 14);
        assert!(std::fs::metadata(tmp.path().join("schedule.pdf")).is_ok());
    }
}
